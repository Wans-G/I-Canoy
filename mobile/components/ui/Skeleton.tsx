import React, { useEffect } from 'react';
import { View, StyleSheet, ViewStyle } from 'react-native';
import Animated, { 
  useSharedValue, 
  withRepeat, 
  withTiming, 
  useAnimatedStyle,
  cancelAnimation
} from 'react-native-reanimated';
import { DimensionValue } from 'react-native';

interface SkeletonProps {
  width?: DimensionValue;
  height?: DimensionValue;
  borderRadius?: number;
  style?: ViewStyle;
  animate?: boolean;
}

const Skeleton: React.FC<SkeletonProps> = ({
  width = '100%',
  height = 20,
  borderRadius = 4,
  style,
  animate = true,
}) => {
  const opacity = useSharedValue(0.5);

  useEffect(() => {
    if (animate) {
      opacity.value = withRepeat(
        withTiming(1, { duration: 800 }),
        -1,
        true
      );
    }
    
    return () => {
      cancelAnimation(opacity);
    };
  }, [animate, opacity]);

  const animatedStyle = useAnimatedStyle(() => ({
    opacity: opacity.value,
  }));

  return (
    <Animated.View
      style={[
        styles.skeleton,
        animatedStyle,
        { width, height, borderRadius },
        style,
      ]}
    />
  );
};

const styles = StyleSheet.create({
  skeleton: {
    backgroundColor: '#E1E9EE',
  },
});

export default Skeleton; 