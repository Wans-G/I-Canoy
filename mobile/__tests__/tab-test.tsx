import React from "react";
import { render } from "@testing-library/react-native";
import { Platform, Text } from "react-native";

// Mocks for custom components
jest.mock("@/components/HapticTab", () => ({
  HapticTab: jest.fn(() => null),
}));

jest.mock("@/components/ui/IconSymbol", () => {
  const React = require("react");
  const { Text } = require("react-native");
  return {
    IconSymbol: jest.fn(({ name, size, color }) => (
      <Text testID={`icon-${name}`}>{`${name}:${size}:${color}`}</Text>
    )),
  };
});

jest.mock("@/components/ui/TabBarBackground", () => {
  const React = require("react");
  return jest.fn(() => <></>);
});

jest.mock("@/hooks/useColorScheme", () => ({
  useColorScheme: () => "light",
}));

jest.mock("@/constants/Colors", () => ({
  Colors: {
    light: { tint: "#123456" },
    dark: { tint: "#654321" },
  },
}));

// Enhanced Tabs mock that invokes tabBarIcon
jest.mock("expo-router", () => {
  const React = require("react");
  const { Text } = require("react-native");

  return {
    Tabs: Object.assign(({ children }: any) => <>{children}</>, {
      Screen: ({ name, options }: any) => {
        const Icon = options?.tabBarIcon?.({ color: "mock-color" }) || null;
        return (
          <>
            <Text>{name}</Text>
            {Icon}
          </>
        );
      },
    }),
  };
});

// Import target component
import TabLayout from "../app/(tabs)/_layout";
import { IconSymbol } from "@/components/ui/IconSymbol";

describe("TabLayout", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders all tab screens", () => {
    const { getByText } = render(<TabLayout />);
    ["home", "store", "social", "profile"].forEach((screen) => {
      expect(getByText(screen)).toBeTruthy();
    });
  });

  it("renders icons with correct props", () => {
    render(<TabLayout />);
    expect(IconSymbol).toHaveBeenCalledWith(
      expect.objectContaining({ name: "house.fill", size: 28, color: "mock-color" }),
      {}
    );
    expect(IconSymbol).toHaveBeenCalledWith(
      expect.objectContaining({ name: "bag.fill", size: 28, color: "mock-color" }),
      {}
    );
    expect(IconSymbol).toHaveBeenCalledWith(
      expect.objectContaining({ name: "message", size: 28, color: "mock-color" }),
      {}
    );
    expect(IconSymbol).toHaveBeenCalledWith(
      expect.objectContaining({ name: "person", size: 28, color: "mock-color" }),
      {}
    );
  });

  it("uses iOS-specific tabBarStyle", () => {
    jest.spyOn(Platform, "select").mockImplementation((options) => options.ios);
    const { toJSON } = render(<TabLayout />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("uses default tabBarStyle on Android", () => {
    jest.spyOn(Platform, "select").mockImplementation((options) => options.default);
    const { toJSON } = render(<TabLayout />);
    expect(toJSON()).toMatchSnapshot();
  });
});
