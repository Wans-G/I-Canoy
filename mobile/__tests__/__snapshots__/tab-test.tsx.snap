// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabLayout uses default tabBarStyle on Android 1`] = `
[
  <Text>
    home
  </Text>,
  <Text
    testID="icon-house.fill"
  >
    house.fill:28:mock-color
  </Text>,
  <Text>
    store
  </Text>,
  <Text
    testID="icon-bag.fill"
  >
    bag.fill:28:mock-color
  </Text>,
  <Text>
    social
  </Text>,
  <Text
    testID="icon-message"
  >
    message:28:mock-color
  </Text>,
  <Text>
    profile
  </Text>,
  <Text
    testID="icon-person"
  >
    person:28:mock-color
  </Text>,
]
`;

exports[`TabLayout uses iOS-specific tabBarStyle 1`] = `
[
  <Text>
    home
  </Text>,
  <Text
    testID="icon-house.fill"
  >
    house.fill:28:mock-color
  </Text>,
  <Text>
    store
  </Text>,
  <Text
    testID="icon-bag.fill"
  >
    bag.fill:28:mock-color
  </Text>,
  <Text>
    social
  </Text>,
  <Text
    testID="icon-message"
  >
    message:28:mock-color
  </Text>,
  <Text>
    profile
  </Text>,
  <Text
    testID="icon-person"
  >
    person:28:mock-color
  </Text>,
]
`;
